### PR TITLE
Strip extra null characters in socketAddress

### PR DESCRIPTION
**Pull request by dlech**
Fixes <https://bugzilla.xamarin.com/show_bug.cgi?id=35004>.

Also remove unnecessary copying of array.

### CODE REVIEW

File: mcs/class/Mono.Posix/Mono.Unix/UnixEndPoint.cs [hunk 1]
```diff
@@ -82,12 +82,10 @@ public override EndPoint Create (SocketAddress socketAddress)
 				uep.filename = "";
 				return uep;
 			}
-			byte [] bytes = new byte [socketAddress.Size - 2 - 1];
-			for (int i = 0; i < bytes.Length; i++) {
-				bytes [i] = socketAddress [i + 2];
-			}
 
-			string name = Encoding.Default.GetString (bytes);
+			string name = Encoding.Default.GetString (bytes, 2, socketAddress.Size - 2 - 1);
```
Comment (akoeplinger) [inline mcs/class/Mono.Posix/Mono.Unix/UnixEndPoint.cs]: `bytes` doesn't exist here.

File: mcs/class/Mono.Posix/Mono.Unix/UnixEndPoint.cs [hunk 2]
```diff
@@ -82,12 +82,10 @@ public override EndPoint Create (SocketAddress socketAddress)
...
+			// remove any extra null characters
+			name = name.Replace ("\0", "");
```
Comment (akoeplinger) [inline mcs/class/Mono.Posix/Mono.Unix/UnixEndPoint.cs]: this could be `name = name.TrimEnd ('\0');`

**Comment by akoeplinger** [comment]
Can you please add the unit test to https://github.com/mono/mono/tree/master/mcs/class/Mono.Posix/Test/Mono.Unix? (we don't use the `Assert.That ()` NUnit style, so please adapt to the existing style).

**Comment by dlech** [comment]
Thanks for the review. Yes, I will be glad to fix this up.

**Comment by dlech** [comment]
Well, I added some unit tests, but it turns out that my fix doesn't work. I am getting two extra random characters when running the unit test instead of two extra null characters. I'll have to do some more digging to figure it out - unless someone else has any insight.

**Comment by dlech** [comment]
OK. Should be good now. I've added a unit test for the original problem that I was seeing (UnixListenerTest) and a unit test for the method that was actually changed (UnixEndPointTest). Both of these tests fail on mono master. Then I added a separate commit with a new fix that is more robust than the previous one.

**Comment by akoeplinger** [comment]
build

**Comment by akoeplinger** [comment]
Looks good to me! I've triggered a build on the CI (expect a few test cases failing right now, they're unrelated).

Any other comments from people more familiar with the code?

**Comment by dlech** [comment]
The build failed because I was using Path.GetFullPath() in one of the unit tests and the build server has a very long path name where the tests are running. Unix sockets are limited to 108 characters in the path (see `man 7 unix`). This is why it worked locally for me and failed in CI. We have to have an absolute path. I tried it and the test fails if I use a relative path. So, I have replaced Path.GetFullPath() with Path.GetTempFileName(), which should hopefully get a shorter file name in CI.  Please give it another go.

**Comment by akoeplinger** [comment]
build

**Comment by akoeplinger** [comment]
It looks good to me.

@jonpryor can you please take a look as well?

**Comment by akoeplinger** [comment]
@dlech would you mind rebasing this to get rid of the merge conflicts?

**Comment by dlech** [comment]
rebased

**Comment by akoeplinger** [comment]
build

**Comment by akoeplinger** [comment]
LGTM.
